src/tree/gpu_hist/row_partitioner.cu

```diff
@@ -93,23 +93,26 @@ void RowPartitioner::SortPosition(common::Span<bst_node_t> position,
                                  position.size(), stream);
 }
 
-void Reset(int device_idx, common::Span<RowPartitioner::RowIndexT> ridx,
-           common::Span<bst_node_t> position) {
-  CHECK_EQ(ridx.size(), position.size());
-  dh::LaunchN(device_idx, ridx.size(), [=] __device__(size_t idx) {
-    ridx[idx] = idx;
-    position[idx] = 0;
-  });
-}
-
 RowPartitioner::RowPartitioner(int device_idx, size_t num_rows)
-    : device_idx_(device_idx), ridx_a_(num_rows), position_a_(num_rows) {
+    : device_idx_(device_idx) {
   dh::safe_cuda(cudaSetDevice(device_idx_));
-  Reset(device_idx, dh::ToSpan(ridx_a_), dh::ToSpan(position_a_));
+  ridx_a_.resize(num_rows);
+  ridx_b_.resize(num_rows);
+  position_a_.resize(num_rows);
+  position_b_.resize(num_rows);
+  ridx_ = dh::DoubleBuffer<RowIndexT>{&ridx_a_, &ridx_b_};
+  position_ = dh::DoubleBuffer<bst_node_t>{&position_a_, &position_b_};
+  ridx_segments_.emplace_back(Segment(0, num_rows));
+
+  thrust::sequence(
```

Comment:
These two thrust functions cause synchronisation, can you keep the Reset function without reintroducing the bug?